practical-penguins/trivia_tavern/trivia_builder/views.py

```diff
@@ -0,0 +1,124 @@
+        active_trivia_quiz.save()
+        return HttpResponseRedirect(reverse('activequiz', kwargs={'pk': active_trivia_quiz.pk}))
+
+
+class TriviaQuizCreateView(PassRequestToFormViewMixin, LoginRequiredMixin, CreateView):
+    model = TriviaQuiz
+    fields = ['name']
+    template_name = 'trivia_builder/triviaquiz_form.html'
+
+    def post(self, request, *args, **kwargs):
+        quiz_form = TriviaQuizForm(request.POST)
+        TriviaQuestionFormSet = modelformset_factory(TriviaQuestion, form=TriviaQuestionForm)
+        question_formset = TriviaQuestionFormSet(request.POST)
+        new_questions = []
+        if quiz_form.is_valid() and question_formset.is_valid():
+            quiz_form.instance.author = self.request.user
+            quiz = quiz_form.save()
+            for i, question_form in enumerate(question_formset.forms, start=1):
+                question_form.instance.quiz = quiz
+                question_form.instance.question_index = i
+                question = question_form.save()
+                new_questions.append(question)
+            question_formset.save()
+
+            try:
+                with transaction.atomic():
+                    TriviaQuestion.objects.bulk_create(new_questions)
+                    quiz.save()
+
+                    # And notify our users that it worked
+                    messages.success(request, 'You have created your quiz.')
+                    return HttpResponseRedirect(reverse('quiz-list'))
+
+            except IntegrityError:  # If the transaction failed
+                messages.error(request, 'There was an error saving your profile.')
+                return redirect(reverse('quiz-list'))
+        return render(request, self.template_name, {'quiz_form': quiz_form, 'question_formset': question_formset})
+
+    def get(self, request, *args, **kwargs):
+        quiz_form = TriviaQuizForm()
+        TriviaQuestionFormSet = modelformset_factory(TriviaQuestion, form=TriviaQuestionForm, extra=1)
+        question_formset = TriviaQuestionFormSet(queryset=TriviaQuestion.objects.none())
+        return render(request, self.template_name, {'quiz_form': quiz_form, 'question_formset': question_formset})
+
+
+class TriviaQuizUpdateView(LoginRequiredMixin, UserPassesTestMixin, UpdateView):
+    model = TriviaQuiz
+    fields = ['name']
+
+    def form_valid(self, form):
+        form.instance.author = self.request.user
+        return super().form_valid(form)
+
+    def test_func(self):
+        trivia_quiz = self.get_object()
+        if self.request.user == trivia_quiz.author:
+            return True
+        return False
+
+
+class TriviaQuizDeleteView(LoginRequiredMixin, UserPassesTestMixin, DeleteView):
+    model = TriviaQuiz
+    success_url = '/quiz'
+
+    def test_func(self):
+        trivia_quiz = self.get_object()
+        if self.request.user == trivia_quiz.author:
+            return True
+        return False
```

Comment:
overall nice classes, good use of mixins, short methods. I would love to see some docstrings and maybe type anotations just to make it easier to read.

Reply:
Agreed, the classes are easy to read and well organized but could benefit quite a bit from docstrings to summarize their purpose.

I'm not a huge fan of the scattered `test_func()`s though; those should be in a separate `test_views.py` or `test_*.py` file.